File: demo/src/main/java/com/example/OrderProducer.java
package com.example;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import java.util.Random;

/**
 * OrderProducer - Productor de pedidos para el sistema de procesamiento asíncrono
 * Este productor simula la creación de pedidos de una tienda online
 * y los envía a una cola de RabbitMQ para ser procesados
 */
public class OrderProducer {
    // Nombre de la cola donde se enviarán los pedidos
    private final static String QUEUE_NAME = "order_queue";

    public static void main(String[] args) {
        // Configurar la conexión a RabbitMQ
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("localhost");

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {
            
            // Declarar la cola - se creará si no existe
            // Parámetros: nombre, durable, exclusive, autoDelete, argumentos
            channel.queueDeclare(QUEUE_NAME, false, false, false, null);

            // Simular la creación de 10 pedidos
            Random random = new Random();
            String[] productos = {"Laptop", "Mouse", "Teclado", "Monitor", "Audífonos"};
            
            for (int i = 1; i <= 10; i++) {
                // Crear un pedido con formato JSON simple
                String producto = productos[random.nextInt(productos.length)];
                int cantidad = random.nextInt(5) + 1;
                double precio = (random.nextDouble() * 1000) + 50;
                
                String pedido = String.format(
                    "{\"orderId\": %d, \"producto\": \"%s\", \"cantidad\": %d, \"precio\": %.2f}",
                    i, producto, cantidad, precio
                );

                // Publicar el mensaje en la cola
                // Parámetros: exchange, routingKey, props, mensaje en bytes
                channel.basicPublish("", QUEUE_NAME, null, pedido.getBytes("UTF-8"));
                
                System.out.println(" [✓] Pedido enviado: " + pedido);
                
                // Pequeña pausa para simular pedidos en tiempo real
                Thread.sleep(500);
            }

            System.out.println("\n [✓] Todos los pedidos han sido enviados exitosamente!");

        } catch (Exception e) {
            System.err.println("Error al enviar pedidos: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: demoPy/OrderProducer.py
import pika
import json
import random
import time

"""
OrderProducer - Productor de pedidos para el sistema de procesamiento asíncrono
Este productor simula la creación de pedidos de una tienda online
y los envía a una cola de RabbitMQ para ser procesados
"""

def main():
    # Nombre de la cola donde se enviarán los pedidos
    QUEUE_NAME = 'order_queue'
    
    try:
        # Conectar con RabbitMQ en localhost
        connection = pika.BlockingConnection(pika.ConnectionParameters('localhost'))
        channel = connection.channel()

        # Declarar la cola - se creará si no existe
        channel.queue_declare(queue=QUEUE_NAME)

        # Lista de productos para simular pedidos
        productos = ['Laptop', 'Mouse', 'Teclado', 'Monitor', 'Audífonos']

        # Simular la creación de 10 pedidos
        for i in range(1, 11):
            # Crear un pedido con datos aleatorios
            pedido = {
                'orderId': i,
                'producto': random.choice(productos),
                'cantidad': random.randint(1, 5),
                'precio': round(random.uniform(50, 1050), 2)
            }

            # Convertir el pedido a JSON
            mensaje = json.dumps(pedido)

            # Publicar el mensaje en la cola
            # exchange='': usa el exchange por defecto
            # routing_key: nombre de la cola
            channel.basic_publish(
                exchange='',
                routing_key=QUEUE_NAME,
                body=mensaje
            )

            print(f" [✓] Pedido enviado: {mensaje}")

            # Pequeña pausa para simular pedidos en tiempo real
            time.sleep(0.5)

        print("\n [✓] Todos los pedidos han sido enviados exitosamente!")

        # Cerrar la conexión
        connection.close()

    except pika.exceptions.AMQPConnectionError as e:
        print(f"Error: No se pudo conectar a RabbitMQ. ¿Está corriendo en localhost?")
        print(f"Detalles: {e}")
    except Exception as e:
        print(f"Error al enviar pedidos: {e}")

if __name__ == '__main__':
    main()
